### PR TITLE
v2: Propagate panics, add `AsyncHandle`

### PR DESCRIPTION
**Pull request by andybarron**
# Code changes
* Use `std::panic` to propagate panics from the thread pool into the async context, rather than triggering the Rayon panic handler.
* Add `AsyncRayonHandle` type that implements `Future`, which makes the `async-trait` crate unnecessary.
* Bypass Tokio `RecvError`. We control the `Sender`, so it should never be dropped too early.
* Remove `prelude` module.
* Seal `AsyncThreadPool` trait.

# Repository changes
* Update CI settings to support Rust 1.45.0 and up.
* Update Codecov settings.
* Add changelog.
* Update README.

### CODE REVIEW

**Comment by coolreader18** [comment]
Coming from the reddit thread - I don't think you need to wrap tokio's RecvError, since it's only a unit error. There's no information in it, so you might as well a) have a unit error type of your own like SpawnFailed or b) just unwrap it, since if you're catching panics it shouldn't ever fail to send something through the channel.

**Comment by andybarron** [comment]
> Coming from the reddit thread - I don't think you need to wrap tokio's RecvError, since it's only a unit error. There's no information in it, so you might as well a) have a unit error type of your own like SpawnFailed or b) just unwrap it, since if you're catching panics it shouldn't ever fail to send something through the channel.

That's a great point! It could only fail if the sender is dropped, but I'm controlling the sender, so that shouldn't ever happen. I'll just have it return unconditionally then :) Thanks!

File: src/async_handle.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+use tokio::sync::oneshot::error::RecvError;
+use tokio::sync::oneshot::Receiver;
+
+/// Async handle for a blocking task running in a Rayon thread pool.
+///
+/// If the spawned task panics, `poll()` will propagate the panic.
+#[must_use]
+#[pin_project]
+#[derive(Debug)]
+pub struct AsyncHandle<T> {
```
Comment (Darksonn) [inline src/async_handle.rs]: I would probably call this a rayon handle. Also, you shouldn't need to pin a oneshot channel as it is already Unpin.

Comment (andybarron) [reply src/async_handle.rs]: Good call to reference `Rayon` in the name. I went with `AsyncRayonHandle` to be extra clear.

Comment (andybarron) [reply src/async_handle.rs]: Also, thanks for the pinning tip - I am new to the concept of pinning so I was overcomplicating it :)

File: src/async_handle.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        })
+        .unwrap_err();
+
+        let (tx, rx) = channel::<thread::Result<()>>();
+        let handle = AsyncRayonHandle { rx };
+        tx.send(Err(panic_err)).unwrap();
+        handle.await;
+    }
+
+    #[tokio::test(flavor = "current_thread")]
```
Comment (Darksonn) [inline src/async_handle.rs]: The current thread runtime is default for tests.

File: Cargo.toml [hunk 3]
```diff
@@ -1,17 +1,17 @@
...
+license = "MIT"
+readme = "README.md"
 repository = "https://github.com/andybarron/tokio-rayon"
 keywords = ["tokio", "rayon", "async", "thread-pool"]
 categories = ["asynchronous", "concurrency", "rust-patterns"]
 
 [dependencies]
 tokio = { version = "^1.4.0", default-features = false, features = ["sync"] }
-async-trait = "^0.1.48"
 rayon = "^1.5.0"
```
Comment (Darksonn) [inline Cargo.toml]: The ^ is redundant. That's the default mode of versions.

File: src/global.rs [hunk 4]
```diff
@@ -1,46 +1,51 @@
...
 /// produciing a future that resolves with the function's return value.
 ///
-/// # Errors
-/// Forwards Tokio's [`RecvError`](tokio::sync::oneshot::error::RecvError), i.e. if the channel is closed.
-pub async fn spawn_async<F, R>(func: F) -> Result<R, RecvError>
+/// # Panics
+/// If the task function panics, the panic will be propagated through the
+/// returned future. Thie will NOT trigger the Rayon thread pool's panic
+/// handler.
+pub fn spawn_async<F, R>(func: F) -> AsyncRayonHandle<R>
```
Comment (Darksonn) [inline src/global.rs]: I'm not sure the `spawn_async` name is great. You aren't spawning an async task here.  Maybe just call it `spawn`?

File: src/global.rs [hunk 5]
```diff
@@ -1,46 +1,51 @@
...
 /// Runs a function on the global Rayon thread pool with LIFO priority,
 /// produciing a future that resolves with the function's return value.
 ///
-/// # Errors
-/// Forwards Tokio's [`RecvError`](tokio::sync::oneshot::error::RecvError), i.e. if the channel is closed.
-pub async fn spawn_async<F, R>(func: F) -> Result<R, RecvError>
+/// # Panics
+/// If the task function panics, the panic will be propagated through the
+/// returned future. Thie will NOT trigger the Rayon thread pool's panic
+/// handler.
```
Comment (Darksonn) [inline src/global.rs:13]: Rayon's panic handler can actually be triggered: if the receivers is dropped before it finishes, the return value is dropped in the rayon thread, and that destructor may panic.

File: src/async_handle.rs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    #[should_panic]
```
Comment (Darksonn) [inline src/async_handle.rs]: This should include the error string, no?

**Review by Darksonn — COMMENTED** [review]
You don't have any panic tests that actually call the spawn method. They all just construct the handle directly.

File: src/global.rs (hunk 4 above)
Comment (andybarron) [reply src/global.rs]: `spawn` on the `AsyncThreadPool` trait conflicts with `spawn` on `rayon::ThreadPool`, and I wanted to use the same name for the function like Rayon does. It's an async wrapper around `spawn` I guess?

File: src/async_handle.rs (hunk 6 above)
Comment (andybarron) [reply src/async_handle.rs]: Good catch!

File: src/global.rs (hunk 5 above)
Comment (andybarron) [reply src/global.rs:13]: Oh yeah - if you dropped the `AsyncHandle`, that would close the channel, and the return value of `func` would be dropped.

That feels like enough of an edge case that we could probably just document it, right? I can't think of any code solutions other than swallowing the destructor panic, which is obviously not great.

File: src/global.rs (hunk 4 above)
Comment (andybarron) [reply src/global.rs]: The more I think about it, I do like `spawn` better, and the trait methods can just be named differently :)

**Comment by andybarron** [comment]
> You don't have any panic tests that actually call the spawn method. They all just construct the handle directly.

Fixed!

File: src/global.rs (hunk 5 above)
Comment (andybarron) [reply src/global.rs:13]: Added a blurb about that scenario to the docs!

**Comment by andybarron** [comment]
@Darksonn @coolreader18 Thanks so much for the feedback! Super helpful :) I'm going to go ahead and merge & release. If there's anything else you notice, feel free to file an issue, and I'll take a look at it :+1:

**Pull request merged**